src/react/hooks/use-presence-listener.ts

```diff
@@ -0,0 +1,173 @@
+/**
+ * The interval between retries when fetching presence data.
+ */
+const PRESENCE_GET_RETRY_INTERVAL_MS = 1500;
+
+/**
+ * The maximum interval between retries when fetching presence data.
+ */
+const PRESENCE_GET_RETRY_MAX_INTERVAL_MS = 30000;
+
+/**
+ * The maximum number of retries when fetching presence data with {@link Presence.get}.
+ */
+const PRESENCE_GET_MAX_RETRIES = 5;
+
+/**
+ * The options for the {@link usePresenceListener} hook.
+ */
+export interface UsePresenceListenerParams extends StatusParams, Listenable<PresenceListener> {
+  /**
+   * The listener to be called when the presence state changes.
+   */
+  listener?: PresenceListener;
+}
+
+export interface UsePresenceListenerResponse extends ChatStatusResponse {
+  /**
+   * The current state of the presence set, observed as a whole change, and only emitted while presence is not syncing.
+   */
+  readonly presenceData: PresenceMember[];
+
+  /**
+   * Provides access to the underlying {@link Presence} instance of the room.
+   */
+  readonly presence: Presence;
+}
+
+/**
+ * A hook that provides access to the {@link Presence} instance in the room and the current presence state.
+ * It will use the instance belonging to the room in the nearest {@link RoomProvider} in the component tree.
+ * On calling, the hook will subscribe to the presence state of the room and update the state accordingly.
+ *
+ * @param params - Allows the registering of optional callbacks.
+ * @returns UsePresenceResponse - An object containing the {@link Presence} instance and the current presence state.
+ */
+export const usePresenceListener = (params?: UsePresenceListenerParams): UsePresenceListenerResponse => {
+  const { currentStatus: connectionStatus, error: connectionError } = useChatConnection({
+    onStatusChange: params?.onConnectionStatusChange,
+  });
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const receivedEventNumber = useRef(0);
+  const triggeredEventNumber = useRef(0);
+  const retryTimeout = useRef<NodeJS.Timeout>();
+  const numRetries = useRef(0);
+  const latestPresentData = useRef<PresenceMember[]>([]);
+  const [presenceData, setPresenceData] = useState<PresenceMember[]>([]);
+
+  useEffect(() => {
+    // handle the received event, ensuring we only process and return the latest presence data.
+    const handleReceivedEvent = () => {
+      receivedEventNumber.current += 1;
+
+      // clear the previous retry if we have received a new event
+      if (retryTimeout.current) {
+        clearTimeout(retryTimeout.current);
+        retryTimeout.current = undefined;
+        numRetries.current = 0;
+      }
+
+      // attempt to get the presence data
+      getAndSetState(receivedEventNumber.current);
+    };
+
+    const getAndSetState = (eventNumber: number) => {
+      room.presence
+        .get({ waitForSync: true })
+        .then((presenceMembers) => {
+          // clear the retry now we have resolved
```

Comment:
do we want all this retry stuff in react only or do we think it's useful in the library too?

Reply:
Andy mentioned this [above](https://github.com/ably/ably-chat-js/pull/324#discussion_r1720037282), I've added a ticket to go back and this to the core SDK :)